net: coap_client: allow application to add block2 option to requests

mrodgers-witekio opened the pull request:
As noted by @SeppoTakalo in the discussion around https://github.com/zephyrproject-rtos/zephyr/issues/76089, if an application adds a Block2 option to a `coap_client` request, this results in duplicate Block2 options in every request of a blockwise transfer after the first one, since the CoAP client appends its own Block2 option with updated NUM and SZX fields.

This PR allows an application to add a Block2 option to an initial request for a resource, which will be ignored by the CoAP client for all subsequent blockwise requests in favour of the updated Block2 option populated inside the client.

I've also added a helper function to correctly populate the Block2 option to send in the initial request, so that the application developer doesn't need to know the details of the Block2 option format, and updated the CoAP client documentation to mention that CoAP options can be added by the application.

mrodgers-witekio committed:
net: coap_client: allow application to add block2 option to requests

Allow an application to add a Block2 option to an initial request for a
resource. For any subsequent requests as part of a blockwise transfer,
drop the application-added Block2 option since the coap_client must
append a Block2 option with updated NUM and SZX fields based on the
server response.

Signed-off-by: Matt Rodgers <mrodgers@witekio.com>

SeppoTakalo commented on doc/connectivity/networking/api/coap_client.rst:
```diff
@@ -79,6 +79,33 @@ The following is an example of a very simple response handling function:
+
+    static struct coap_client;
+    struct coap_client_request req = { 0 };
+    struct coap_client_option block2_option;
```
Just a thought, instead of providing an initializer function for such a static structure, can we just provide a macro for it?

```suggestion

// In coap_client.h
#define COAP_OPTION_BLOCK2_INIT (struct coap_client_option){.code = COAP_OPTION_BLOCK2, .len = 1, .value[0] = coap_bytes_to_block_size(CONFIG_COAP_CLIENT_BLOCK_SIZE)}

// In app code
    struct coap_client_option block2_option = COAP_OPTION_BLOCK2_INIT;
```

SeppoTakalo commented on include/zephyr/net/coap_client.h:
```diff
@@ -159,6 +159,28 @@ int coap_client_req(struct coap_client *client, int sock, const struct sockaddr
+ *
+ * @param option CoAP client option structure to populate
+ */
+static inline void coap_client_option_populate_block2(struct coap_client_option *option)
+{
+	if (option == NULL) {
+		return;
+	}
```
Second option would be that this function would just return the structure.
I think playing with the pointer is unnecessary

```suggestion
static inline struct coap_client_option coap_client_option_block2()
{
        return (struct coap_client_option) { .code = COAP_OPTION_BLOCK2, ...
```

SeppoTakalo reviewed (dismissed):
I agree with the functional change.

I'm just not that liking the provided initializer function. However, that is just an opinion. Feel free to ignore the suggestion. I'll approve anyway.

mrodgers-witekio commented:
> I agree with the functional change.
> 
> I'm just not that liking the provided initializer function. However, that is just an opinion. Feel free to ignore the suggestion. I'll approve anyway.

I considered the macro initializer approach, however the `coap_bytes_to_block_size` function isn't evaluated to a compile time constant. So:

```
struct coap_client_option block2 = COAP_OPTION_BLOCK2_INIT;
```
inside a function works, but 
```
static struct coap_client_option block2 = COAP_OPTION_BLOCK2_INIT;
```
does not - it gives `error: initializer element is not constant`

The convention elsewhere in Zephyr seems to be that these kind of initializer macros are used for things that can be statically defined at compile time (e.g. `Z_SEM_INITIALIZER` and similar), so I don't know if this would be confusing.

I like the "return a struct" approach though, more than likely it'll compile to the same thing as the pointer version anyway.

I'll see if there's any strong opinions from other reviewers before I change anything.

SeppoTakalo commented:
> The convention elsewhere in Zephyr seems to be that these kind of initializer macros are used for things that can be statically defined at compile time (e.g. `Z_SEM_INITIALIZER` and similar), so I don't know if this would be confusing.
> 
> I like the "return a struct" approach though, more than likely it'll compile to the same thing as the pointer version anyway.

Yes, you are right. If we provide macro, it should be static.

Also, the lifetime need to be considered here. During the execution of CoAP client, it seems that the `struct coap_client_request` is shallow copied into the internal request. So the `struct coap_client_options` pointer should stay valid during the whole execution time. If we construct that from the stack frame, we might end up putting garbage options into the query of the second block.

mrodgers-witekio commented:
- Rebased now that CI should be passing again
- Updated `coap_client_option_populate_block2` to instead be `coap_client_option_initial_block2`, returning a struct instead of passing in via pointer
- Updated documentation to show that CoAP options must remain valid for the whole execution of the request

SeppoTakalo reviewed (approved):
Looks good
